token-swap/program/src/curve/fees.rs

```diff
@@ -1,51 +1,77 @@
 
+use crate::error::SwapError;
+use arrayref::{array_mut_ref, array_ref, array_refs, mut_array_refs};
 use solana_program::{
     program_error::ProgramError,
     program_pack::{IsInitialized, Pack, Sealed},
 };
-
-use crate::curve::calculator::{calculate_fee, CurveCalculator, DynPack, SwapWithoutFeesResult};
-use arrayref::{array_mut_ref, array_ref, array_refs, mut_array_refs};
 use std::convert::TryFrom;
 
-/// FlatCurve struct implementing CurveCalculator
+/// Encapsulates all fee information and calculations for swap operations
 #[derive(Clone, Debug, Default, PartialEq)]
-pub struct FlatCurve {
-    /// Fee numerator
+pub struct Fees {
+    /// Trade fees are extra token amounts that are held inside the token
+    /// accounts during a trade, making the value of liquidity tokens rise.
+    /// Trade fee numerator
     pub trade_fee_numerator: u64,
-    /// Fee denominator
+    /// Trade fee denominator
     pub trade_fee_denominator: u64,
+
+    /// Owner trading fees are extra token amounts that are held inside the token
+    /// accounts during a trade, with the equivalent in pool tokens minted to
+    /// the owner of the program.
     /// Owner trade fee numerator
     pub owner_trade_fee_numerator: u64,
     /// Owner trade fee denominator
     pub owner_trade_fee_denominator: u64,
+
+    /// Owner withdraw fees are extra liquidity pool token amounts that are
+    /// sent to the owner on every withdrawal.
     /// Owner withdraw fee numerator
     pub owner_withdraw_fee_numerator: u64,
     /// Owner withdraw fee denominator
     pub owner_withdraw_fee_denominator: u64,
+
+    /// Host fees are a proportion of the owner trading fees, sent to an
+    /// extra account provided during the trade.
     /// Host trading fee numerator
     pub host_fee_numerator: u64,
     /// Host trading fee denominator
     pub host_fee_denominator: u64,
 }
 
-impl CurveCalculator for FlatCurve {
-    /// Flat curve swap always returns 1:1 (minus fee)
-    fn swap_without_fees(
-        &self,
-        source_amount: u128,
-        _swap_source_amount: u128,
-        _swap_destination_amount: u128,
-    ) -> Option<SwapWithoutFeesResult> {
-        Some(SwapWithoutFeesResult {
-            source_amount_swapped: source_amount,
-            destination_amount_swapped: source_amount,
-        })
+/// Helper function for calculating swap fee
+pub fn calculate_fee(
+    token_amount: u128,
+    fee_numerator: u128,
+    fee_denominator: u128,
+) -> Option<u128> {
+    if fee_numerator == 0 || token_amount == 0 {
+        Some(0)
+    } else {
+        let fee = token_amount
+            .checked_mul(fee_numerator)?
+            .checked_div(fee_denominator)?;
+        if fee == 0 {
+            Some(1) // minimum fee of one token
+        } else {
+            Some(fee)
+        }
+    }
+}
+
+fn validate_fraction(numerator: u64, denominator: u64) -> Result<(), SwapError> {
+    if numerator > denominator {
```

Comment:
Can you think of a case where fees should be 100%?

Reply:
Good point, most of the other logic would break because of 0s at that point anyway